zcash_primitives/src/note_encryption.rs

```diff
@@ -0,0 +1,1079 @@
+        let pa = PaymentAddress { diversifier, pk_d };
+
+        // Construct the value commitment for the proof instance
+        let value = 100;
+        let value_commitment = ValueCommitment::<Bls12> {
+            value,
+            randomness: Fs::rand(&mut rng),
+        };
+        let cv = value_commitment.cm(&JUBJUB).into();
+
+        let note = pa.create_note(value, Fs::rand(&mut rng), &JUBJUB).unwrap();
+        let cmu = note.cm(&JUBJUB);
+
+        let ovk = OutgoingViewingKey([0; 32]);
+        let ne = SaplingNoteEncryption::new(ovk, note, pa, Memo([0; 512]));
+        let epk = ne.epk();
+        let enc_ciphertext = ne.encrypt_note_plaintext();
+        let out_ciphertext = ne.encrypt_outgoing_plaintext(&cv, &cmu);
+
+        assert!(try_sapling_note_decryption(&ivk, epk, &cmu, &enc_ciphertext).is_some());
+        assert!(try_sapling_compact_note_decryption(
+            &ivk,
+            epk,
+            &cmu,
+            &enc_ciphertext[..COMPACT_NOTE_SIZE]
+        )
+        .is_some());
+        assert!(try_sapling_output_recovery(
+            &ovk,
+            &cv,
+            &cmu,
+            &epk,
+            &enc_ciphertext,
+            &out_ciphertext
+        )
+        .is_some());
+
+        (
+            ovk,
+            ivk,
+            cv,
+            cmu,
+            epk.clone(),
+            enc_ciphertext,
+            out_ciphertext,
+        )
+    }
+
+    #[test]
+    fn decryption_with_invalid_ivk() {
+        let mut rng = thread_rng();
+
+        let (_, _, _, cmu, epk, enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        assert_eq!(
+            try_sapling_note_decryption(&Fs::rand(&mut rng), &epk, &cmu, &enc_ciphertext),
+            None
+        );
+    }
+
+    #[test]
+    fn decryption_with_invalid_epk() {
+        let mut rng = thread_rng();
+
+        let (_, ivk, _, cmu, _, enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        assert_eq!(
+            try_sapling_note_decryption(
+                &ivk,
+                &edwards::Point::<Bls12, _>::rand(&mut rng, &JUBJUB).mul_by_cofactor(&JUBJUB),
+                &cmu,
+                &enc_ciphertext
+            ),
+            None
+        );
+    }
+
+    #[test]
+    fn decryption_with_invalid_cmu() {
+        let mut rng = thread_rng();
+
+        let (_, ivk, _, _, epk, enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        assert_eq!(
+            try_sapling_note_decryption(&ivk, &epk, &Fr::rand(&mut rng), &enc_ciphertext),
+            None
+        );
+    }
+
+    #[test]
+    fn decryption_with_invalid_tag() {
+        let mut rng = thread_rng();
+
+        let (_, ivk, _, cmu, epk, mut enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        enc_ciphertext[ENC_CIPHERTEXT_SIZE - 1] ^= 0xff;
+        assert_eq!(
+            try_sapling_note_decryption(&ivk, &epk, &cmu, &enc_ciphertext),
+            None
+        );
+    }
```

Comment:
There's a missing failure case for all three decryption types: the diversifier may be invalid so that g<sub>d</sub> can't be computed. `random_enc_ciphertext` will never create a ciphertext that triggers this case because it always uses a fixed (all-zeros) diversifier.